src/primitive/Primitive.ts

```diff
@@ -0,0 +1,216 @@
+            this.position.copy(newPos);
+
+            // if the position changed, update the object in communication
+            if (this.position.y === oldPos.y) return;
+            DIVECommunication.get(this.userData.id)?.PerformAction('UPDATE_OBJECT', { id: this.userData.id, position: this.position, rotation: this.rotation, scale: this.scale });
+        }
+    }
+
+    public onMove(): void {
+        DIVECommunication.get(this.userData.id)?.PerformAction('UPDATE_OBJECT', { id: this.userData.id, position: this.position, rotation: this.rotation, scale: this.scale });
+    }
+
+    public onSelect(): void {
+        DIVECommunication.get(this.userData.id)?.PerformAction('SELECT_OBJECT', { id: this.userData.id });
+    }
+
+    public onDeselect(): void {
+        DIVECommunication.get(this.userData.id)?.PerformAction('DESELECT_OBJECT', { id: this.userData.id });
+    }
+
+    private assembleGeometry(geometry: COMGeometry): BufferGeometry {
+        switch (geometry.name) {
+            case 'cylinder':
+                return this.createCylinderGeometry(geometry);
+            case 'sphere':
+                return this.createSphereGeometry(geometry);
+            case 'pyramid':
+                return this.createPyramidGeometry(geometry);
+            case 'box':
+                return this.createBoxGeometry(geometry);
+            case 'cone':
+                return this.createConeGeometry(geometry);
+            case 'wall':
+                return this.createWallGeometry(geometry);
+            case 'plane':
+                return this.createPlaneGeometry(geometry);
+            default:
+                return new BufferGeometry();
+        }
+    }
+
+    private createCylinderGeometry(geometry: COMGeometry): BufferGeometry {
+        const geo = new CylinderGeometry(geometry.width * 2, geometry.width * 2, geometry.height, 64);
+        return geo;
+    }
+
+    private createSphereGeometry(geometry: COMGeometry): BufferGeometry {
+        const geo = new SphereGeometry(geometry.width * 2, 64);
+        return geo;
+    }
+
+    private createPyramidGeometry(geometry: COMGeometry): BufferGeometry {
+        const geo = new BufferGeometry();
+        const { width, height, depth } = geometry;
+        geo.setAttribute('position', new Float32BufferAttribute([
+            width / 2, 0, depth / 2, // right back
+            width / 2, 0, -depth / 2, // right front
+            -width / 2, 0, -depth / 2, // left front
+            -width / 2, 0, depth / 2, // left back
+            0, height, 0, // top
+        ], 3));
+        geo.setIndex(new Uint32BufferAttribute([
+            1, 0, 4,
+            2, 1, 4,
+            3, 2, 4,
+            3, 0, 4,
+            0, 1, 2,
+            0, 2, 3,
+        ], 1));
+        return geo;
+    }
+
+    private createBoxGeometry(geometry: COMGeometry): BufferGeometry {
+        const geo = new BoxGeometry(geometry.width, geometry.height, geometry.depth);
```

Comment:
Approved, just a side note I guess some of the lasts createXZYGeometry functions can return directly, bypassing the constant creation as the output is clear enough 👍 .

Reply:
Yep, true, I will change this! :)

Reply:
Please check again, thank you in advance!

Reply:
btw I just changed that given approvals are revoked then pushing something after the approval